### ISSUE TITLE
Need help: High Std of Reward

### ISSUE THREAD

**Issue by JevMar** [Unity-Technologies/ml-agents#960]
Hi, I'm having some difficulties while training agents to navigate in an environment with obstacles with  Curriculum learning. However, the problem is more about not being able to switch lessons due to high fluctuation of the Cumulative reward.

At the moment, even though the agents can learn the first level pretty quickly (using 12 to train), I keep facing the same issue - the standard deviation(Std) of the reward stays pretty high even after long training:
![high std of reward](https://user-images.githubusercontent.com/31615473/42425203-f900ee2c-8319-11e8-84b4-78815afb988c.PNG)

Below are the training results where reward threshold = 0.850
![curriculum_0 850_threshold](https://user-images.githubusercontent.com/31615473/42425207-17b081ac-831a-11e8-877c-50b593d3bae2.PNG)

With reward threshold = 0.800 the lessons switch, however, after a certain point the Mean reward decreases:
![curriculum_0 8_threshold](https://user-images.githubusercontent.com/31615473/42425228-80dd0b0a-831a-11e8-8a0a-5cd744864346.PNG)

**Environment setup:**
Goal: The agent must get to target position.
Agents: The environment contains 12 identical agents, which are all contained within their own training areas and are linked to a single brain.
Rewards:
    +1.0 for reaching the target.
    -1.0 for hitting a wall.
    0.0 if the time runs out.

Observation space:
    X and Z positions of the agent relative to the target.
    X and Z rotations of the agent.
    3 sensors that can hit walls and output distance in range of [0;1], where 0 - no hits, 1 - collision with a wall.

Action space (Discrete):
    Move forward per 0.1 unit.
    Rotate left or right per 10 degrees.

The agent resets if:
    collides with a wall.
    reaches the target.
    reaches limit of 500 steps.

The academy has max_steps = 5000

Curriculum JSON:
{
    "measure" : "reward",
    "thresholds" : [0.8, 0.8, 0.8, 0.8],
    "min_lesson_length" : 1,
    "signal_smoothing" : true,
    "parameters" :
    {
        "EnvLevel" : [0.0, 1.0, 2.0, 3.0, 4.0]
    }
}

I have run a great number of tests to figure out which observations work and how to tune the hyperparameters for best results. Currently using
batch_size: 256
beta: 0.05
buffer_size: 2560
hidden_units: 96
The rest are same as per default preset.

Can anyone give any suggestion on how could I improve this?

**Comment by scotthovestadt** [comment]
@JevMar Have you tried:

1- `use_curiosity` can increase initial training time but I've found can generate more reproducible results, since the agent will prioritize figuring out it's environment instead of striving for a local minimum. You will probably also need to adjust your other hyperparameters. See the Pyramid config for an example.

2- I would avoid a penalty for hitting a wall. Instead, just end the lesson and give more of a reward for reaching the target quickly. Ensure that if they reach the wall in a reasonable amount of time they can get the full reward (scale this as appropriate). This will teach them to avoid walls to be faster at reaching the target. You can always add a penalty for touching a wall in a later lesson and see if that helps.

**Comment by JevMar** [comment]
@scotthovestadt Thanks for the suggestions, will give both a shoot. Although, regarding the 2nd one - would you still suggest to use +1.0 as the full reward and scale it down overtime? I have read that best practice is to keep rewards in range of -1 to +1, however, I have also seen quite few examples where maximum reward is way higher, like+20, etc.

**Comment by scotthovestadt** [comment]
@JevMar I have found normalization to be absolutely key. I take great care to never pass any values outside of -1/1 or 0/1. I also never reward the agent more than 1 or -1 in a single step.

What I am currently doing for "reaching a goal" type rewards is something like this:
`AddReward(Mathf.Min(2 * Mathf.Abs(1f - ((float) GetStepCount() / (float) agentParameters.maxStep)), 1f))`

Basically, the agent will get the full reward of 1 if he reaches the goal in the first 50% of max steps. After that, his reward will scale down linearly.

**Comment by JevMar** [comment]
@scotthovestadt This approach could be a way out here, though, my concern is that with more difficult environment (one with higher number of obstacles, etc.) the agent will need more time to reach the target, therefore, that has to be scaled appropriately as well.

I do have one more question - if an agent has max_steps set to 500, how exactly chaining the Academy max_steps from 5k to 10k (or 20k) would affect the training itself?

**Comment by scotthovestadt** [comment]
@JevMar I'm not sure myself what the best practice for max steps is. I've just been setting my max steps per agent and academy to a fairly high number (the same number). It's something I need to work on understanding more.

In my experience the scaling isn't that hard because you don't want to over-train on early courses-- only teach concepts to the AI. My early courses are designed to reward the agent for doing a decent job and then they advance. I'm currently advancing them when the reward threshold is 0.75. I introduce new concepts like random spawn position, random goal position, obstacles, etc. I don't need them to do an amazing job in these early courses, it's just teaching them to accomplish the goal reliably in a reasonable time period.

I've actually found that at this step of the training process my agents behave rather conservatively, not even traveling at max speed. What I have also found is that they have a pretty great final success rate and when watching them their failures at this stage make sense most of the time. I think the reward system is teaching them that accomplishing the goal reliably is better than shaving off milliseconds.

Eventually I graduate my agent to a final course that has everything: random spawn, random size, random goal position, random obstacles, multiple goals, moving goals, etc. I consider the previous training just setup for them to begin learning on this course. At this point I take great care to tune the rewards.

I've found in my situation that a limited time period with multiple goals and a flat reward of 1 for each goal they get during the time period encourages them in the most optimal way at this stage. At earlier stages this didn't work because they would get dragged into a local optimum of just going in expanding circles to collect a goal or two.

After a while of training on this course they still do excellent when brought back to a single goal course.

**Comment by scotthovestadt** [comment]
@JevMar Note also I control almost everything via reset parameters, so that scaled reward that they get can be adjusted if necessary. I wrote a script to make writing the lessons a bit easier by adding default parameters to the JSON schema and pushing the entire lesson into a single object instead of arrays of floats. I'd be happy to share the script if you're interested.

**Comment by JevMar** [comment]
@scotthovestadt I'm definitely interested! 
Also, if your agent has a fairly high number set as max_steps, do you reset both, agent and environment, after the goal is reached? I'm currently resetting both, however, I'm wondering whether it would be a better approach if the agent gets a second goal instead (meaning only the environment resets).

**Comment by scotthovestadt** [comment]
Here is an example of the lesson file format I am using. I find it easier to edit and understand:

```
{
    "measure": "reward",
    "min_lesson_length": 3,
    "signal_smoothing": true,
    "default_parameters": {
        "randomize_spawn_position": 0,
        "spawn_position_x": 0.5,
        "spawn_position_z": 0.5,
        "ground_size_x": 100,
        "ground_size_z": 100,
        "num_waypoints": 1,
        "waypoint_collider_radius": 3,
        "waypoint_min_distance": 10,
        "waypoint_min_x": 0,
        "waypoint_max_x": 1,
        "waypoint_min_z": 0,
        "waypoint_max_z": 1,
        "num_obstacles": 0,
        "obstacle_size": 5,
        "obstacle_min_x": 0,
        "obstacle_max_x": 1,
        "obstacle_min_z": 0,
        "obstacle_max_z": 1,
        "reward_waypoint_reached": 0,
        "reward_success_by_timer": 0,
        "penalty_fall_off_edge": 0,
        "penalty_collision": 0,
        "penalty_existential": 0
    },
    "lessons": [
        {
            "ground_size_x": 10,
            "ground_size_z": 20,
            "waypoint_min_x": 0.5,
            "waypoint_max_x": 0.5,
            "waypoint_min_z": 1,
            "waypoint_max_z": 1,
            "spawn_position_x": 0,
            "spawn_position_z": 0,
            "reward_success_by_timer": 1.5,
            "threshold": 0.75
        },
        {
            "ground_size_x": 10,
            "ground_size_z": 20,
            "waypoint_min_x": 0,
            "waypoint_max_x": 1,
            "waypoint_min_z": 1,
            "waypoint_max_z": 1,
            "spawn_position_x": 0,
            "spawn_position_z": 0,
            "reward_success_by_timer": 1.5,
            "threshold": 0.75
        },
        {
            "ground_size_x": 75,
            "ground_size_z": 75,
            "waypoint_min_x": 0.1,
            "waypoint_max_x": 0.9,
            "waypoint_min_z": 0.75,
            "waypoint_max_z": 0.75,
            "spawn_position_x": 0.25,
            "spawn_position_z": 0.25,
            "reward_success_by_timer": 1.5,
            "threshold": 0.75
        },
        {
            "ground_size_x": 75,
            "ground_size_z": 75,
            "waypoint_min_x": 0.1,
            "waypoint_max_x": 0.9,
            "waypoint_min_z": 0.75,
            "waypoint_max_z": 0.75,
            "randomize_spawn_position": 1,
            "reward_success_by_timer": 1.5,
            "threshold": 0.75
        }
    ]
}
```

Here is the script I wrote to generate it. I probably should have written it in Python but I wanted to be able to be explicit about the schema of the objects so I used TypeScript. It's still a little rough:
https://gist.github.com/scotthovestadt/e8e530a0047873963a38e95556875eac

Here is the plain JavaScript version if you don't use TypeScript:
https://gist.github.com/scotthovestadt/56a663d1a5e788d965087fe078cc3c0f

Drag the file into your directory, change the filename referenced at the top of the file to your lesson filename, and then run learn.py like this:
`node generate-lesson-plan.js && python3 learn.py ...`

All the script does is take these custom parameters and transform them into the parameters Python needs and then write both back to the file.

Let me know if you find it useful. Maybe I'll clean it up and convert it to Python so it's easier to use.

**Comment by scotthovestadt** [comment]
@JevMar I have a configurable number of agents (it generates X environments+environments on start from one that I place manually in the scene and edit) so I only reset individual agents when they finish. The academy does it's own thing and has the same number of max steps. I'm not 100% sure about my configuration here and it's on my todo list to understand more if this is optimal.

**Comment by JevMar** [comment]
@scotthovestadt Huge thank you for all the suggestions and for sharing the script! I'll give an update on my results, if I manage to improve them.

**Comment by JevMar** [comment]
I run few more tests to figure out what is the issue here. Created a new but similar environment with a simplified goal + changed reward system according to @scotthovestadt advice. Tried few variations: target object had same position each time, then it was randomized, finally added few obstacles.
Results: agents learn really fast, however, same fluctuation can be observed.
![image](https://user-images.githubusercontent.com/31615473/42539065-1f9c545a-849a-11e8-854e-82555413039e.png)

Finally, after observing the agent's behavior I realized that these slight drops are happening each time the Academy resets. To test, tried one more training. Green line below represents new training with twice the amount of academy max_steps. Result: drops happen less often.
![image](https://user-images.githubusercontent.com/31615473/42539432-6ca06394-849b-11e8-88d4-0d0a4517e1be.png)

Does this mean I have to reset agents through academy to avoid this behavior?

**Comment by scotthovestadt** [comment]
@JevMar Happy to hear they learn faster. That's been my experience as well with good lesson plans.

I am also curious what the best practice is for max steps when doing lessons plans. I looked at the Wall Jump project and the Academy has 10k steps with the Agent having 2k. I'll probably try a similar ratio with my project and see if there's any difference.

**Comment by mmattar** [comment]
Hey @JevMar / @scotthovestadt - just following up here to check-in on progress and see if there are any additional lingering questions?

**Comment by mmattar** [comment]
Closing this issue for now, but please re-open if there are any additional follow-ups.
